xarray/tests/test_dataarray.py

```diff
@@ -817,6 +817,16 @@ def test_sel_dataarray_datetime(self):
         result = array.sel(delta=slice(array.delta[0], array.delta[-1]))
         assert_equal(result, array)
 
+    def test_sel_float32(self):
+        float_values = [0., 0.111, 0.222, 0.333]
+        coord_values = np.asarray(float_values, dtype='float32')
+        data_values = np.arange(4)
+        array = DataArray(data_values, [('float32_coord', coord_values)])
+        expected = DataArray(data_values[1:3], [('float32_coord',
+                                                 coord_values[1:3])])
+        actual = array.sel(float32_coord=float_values[1:3])
```

Comment:
It would also be good to add a test case where you're only indexing by a scalar.

Reply:
A fix in pandas that I can imagine is to modify the pd.Index to retain the dtype similar to the PandasIndexAdapter in xarray/core/indexing. However, it seems to me that their design it to coerce any float to float64, so I did not feel they would consider this to be a bug. I would try to dig into such solution more if you recommend so.

I added a test case in which the coords are float16 to show that casting to the coords type is required. 

The lowest level in which I can access the coords type is in the indexing/remap_label_indexers where I could get the coords from the data_obj. In the latest code, casting is done just before calling convert_label_indexer

Sorry if I am missing the point, and thank you in advance for any further feedback.